### PR TITLE
[PARK-6489][SQL] add column pruning for Generate

### PR DESCRIPTION
**Pull request by cloud-fan**
This PR takes over https://github.com/apache/spark/pull/5358

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -178,6 +179,10 @@ object ColumnPruning extends Rule[LogicalPlan] {
     case a @ Aggregate(_, _, child) if (child.outputSet -- a.references).nonEmpty =>
       a.copy(child = Project(a.references.toSeq, child))
 
+    // Eliminate attributes that are not needed to calculate the Generate.
+    case g: Generate if !g.join && (g.child.outputSet -- g.references).nonEmpty =>
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Can we also handle the join case when there is a project on top of the generate (i.e. push the project through the generate)?

**Comment by cloud-fan** [comment]
retest this please.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ColumnPruningSuite.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        Generate(Explode('b), true, false, None, 's.string :: Nil,
+          input)).analyze
+    val optimized = Optimize.execute(query)
+
+    val correctAnswer =
+      Project(('s + 1).as("s+1") :: Nil,
+        Generate(Explode('b), false, false, None, 's.string :: Nil,
+          input)).analyze
+
+    comparePlans(optimized, correctAnswer)
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ColumnPruningSuite.scala]: Maybe we can also have a test for `val input = LocalRelation('a.int, 'b.array(StringType))`. Basically, we test turning `join` to false and then we add a project as the child of `Generate`.

**Comment by yhuai** [comment]
LGTM.

**Comment by marmbrus** [comment]
Thanks!  Merging to master and 1.5

**Pull request closed without merging**